### PR TITLE
feat: add `remoteDetectionAlert` doc type

### LINKED ISSUE #228: Add alerts type to schema
We need to add a new data type, Alerts, for CMI.

I don't know exactly what fields this type needs to have, which is (I think) the hard part of this work: figuring out what it needs to be. I think this task is fairly straightforward once we figure that out.

### PR DESCRIPTION
**Pull request by tomasciccola**
This adds the `remoteDetectionAlert` doc type. The `geometry` message is a stub for now, until @gmaclennan implements its decoding conversion to GeoJSON.

The only doubt I have is around `sourceId`. Is there a clear format for this string? Should we add concrete validation for it?

Also, I reused the `tags` type for `metadata` (since they are structurally the same?, basically a `Record<string, any>`). I know that semantically is confusing, but this allows to reuse all the conversion functions regarding tags

should close #228

### CODE REVIEW

**Comment by gmaclennan** [comment]
This depends on #248 (because @comapeo/geometry doesn't export `fromPartial` methods, and it's easier to remove them from here than try to write then for Geometry), so that should be merged first.

The actual changes to add geometry are in https://github.com/digidem/mapeo-schema/pull/234/commits/7733276587d1d57109e502e635f914a7c9a896cc

It is a bit more complicated than I would like because of the need to deal with JSONSchema refs with 3 different libraries that we use (`ajv`, `@json-schema-tools/dereferencer` and `@apidevtools/json-schema-ref-parser`).

Would be nice to replace `@json-schema-tools/dereferencer` with `@apidevtools/json-schema-ref-parser` at some point, because the latter is already a dep of `json-schema-to-typescript` and we need to deal with it for that.

**Comment by EvanHahn** [comment]
I plan to review this next week.

File: schema/remoteDetectionAlert/v1.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      }
+    },
+    "geometry": {
+      "$ref": "http://comapeo.app/schemas/shared/geometry.json"
+    }
+  },
+  "required": [
+    "schemaName",
+    "detectionDateStart",
+    "detectionDateEnd",
```
Comment (EvanHahn) [inline schema/remoteDetectionAlert/v1.json:74]: Intuitively I don't know if this should be required. What if there's an alert that's still active?

**Review by EvanHahn — CHANGES_REQUESTED** [review]
I assume you and Gregor decided that this was a good format for alerts.

There's a bunch of Observation stuff in here that needs to be reverted. I _think_ merging `main` should fix that.

Otherwise mostly LGTM.

File: src/lib/decode-conversions.ts [hunk 2]
```diff
@@ -285,6 +291,30 @@ export const convertTrack: ConvertFunction<'track'> = (message, versionObj) => {
...
+  'remoteDetectionAlert'
+> = (message, versionObj) => {
+  const { common, schemaVersion, ...rest } = message
+  const jsonSchemaCommon = convertCommon(common, versionObj)
+  if (!rest.detectionDateStart) {
+    throw new Error('missing required detectionDateStart')
+  }
+  if (!rest.detectionDateEnd) {
+    throw new Error('missing required detectionDateEnd')
+  }
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/lib/decode-conversions.ts:303]: This value is coming from the ts-proto decoder, which will [always decode this to a valid ISO date string](https://github.com/stephenh/ts-proto/blob/main/src/main.ts#L974), even if the value is the default UNIX zero date.

File: proto/geometry/v1.proto
[the remark this answers is not in the document]
Comment (gmaclennan) [reply proto/geometry/v1.proto:1]: Resolved in https://github.com/digidem/mapeo-schema/pull/234/commits/11fbc82c89437d9dfe2fff2a5700848bb97760fc

File: proto/observation/v1.proto
[the remark this answers is not in the document]
Comment (gmaclennan) [reply proto/observation/v1.proto:1]: Resolved in merge https://github.com/digidem/mapeo-schema/pull/234/commits/512d4e92151624954b5bb6b1bb7b1eb2360dd337

File: src/lib/decode-conversions.ts [hunk 3]
```diff
@@ -76,7 +81,7 @@ export const convertObservation: ConvertFunction<'observation'> = (
   message,
   versionObj
 ) => {
-  const { common, schemaVersion, ...rest } = message
+  const { common, metadata, schemaVersion, ...rest } = message
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/lib/decode-conversions.ts]: Resolved in merge https://github.com/digidem/mapeo-schema/pull/234/commits/512d4e92151624954b5bb6b1bb7b1eb2360dd337

File: src/lib/decode-conversions.ts [hunk 4]
```diff
@@ -453,3 +483,35 @@ function convertTrackPosition(
...
+  position: Observation_1_Metadata_Position
+): Position | undefined {
+  if (
+    typeof position.coords === 'undefined' ||
+    typeof position.timestamp === 'undefined'
+  ) {
+    return undefined
+  }
+  return position as Position
+}
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/lib/decode-conversions.ts]: Resolved in merge https://github.com/digidem/mapeo-schema/pull/234/commits/512d4e92151624954b5bb6b1bb7b1eb2360dd337

File: test/index.test.js [hunk 5]
```diff
@@ -225,6 +226,75 @@ test(`test encoding of wrongly formatted header`, async () => {
...
+    'undefined',
+    'optional networkAvailable prop should be undefined'
+  )
+  assert.equal(
+    typeof decodedDoc.metadata?.positionProvider?.passiveAvailable,
+    'undefined',
+    'optional passiveAvailable prop should be undefined'
+  )
+})
+
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply test/index.test.js]: Resolved in merge https://github.com/digidem/mapeo-schema/pull/234/commits/512d4e92151624954b5bb6b1bb7b1eb2360dd337

File: src/lib/encode-conversions.ts [hunk 6]
```diff
@@ -95,9 +95,6 @@ export const convertObservation: ConvertFunction<'observation'> = (
   mapeoDoc
 ) => {
   const attachments = mapeoDoc.attachments.map(convertAttachment)
-  const metadata: Observation_1_Metadata | undefined = mapeoDoc.metadata && {
-    ...Observation_1_Metadata.fromPartial(mapeoDoc.metadata),
-  }
```
[the remark this answers is not in the document]
Comment (gmaclennan) [reply src/lib/encode-conversions.ts]: Reserved in merge https://github.com/digidem/mapeo-schema/pull/234/commits/512d4e92151624954b5bb6b1bb7b1eb2360dd337

**Review by gmaclennan — APPROVED** [review]
I think I've resolved all of Evan's comments and I've re-reviewed this, and it looks good to merge to me. @EvanHahn this needs a review approval from you to merge.

**Review by gmaclennan — CHANGES_REQUESTED** [review]
I think we should not merge this until we have released the MVP, in case we need bug fixes before MVP (having this merged would complicate bug fixes)

**Comment by tomasciccola** [comment]
> I think we should not merge this until we have released the MVP, in case we need bug fixes before MVP (having this merged would complicate bug fixes)

wouldn't it make sense to have a 'develop' branch where we merge this so we don't have a bunch of dangling open PRs (but maybe there won't be many of this)?

File: proto/remoteDetectionAlert/v1.proto [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+message RemoteDetectionAlert_1 {
+  // **DO NOT CHANGE dataTypeId** generated with `openssl rand -hex 6`
+  option (dataTypeId) = "33a966ee6c90";
+  option (schemaName) = "remoteDetectionAlert";
+
+  Common_1 common = 1;
+
+  google.protobuf.Timestamp detectionDateStart = 2;
+  google.protobuf.Timestamp detectionDateEnd = 3;
```
Comment (EvanHahn) [inline proto/remoteDetectionAlert/v1.proto:18]: Could we add `[(required) = true]` to these?

```suggestion
  google.protobuf.Timestamp detectionDateStart = 2 [(required) = true];
  google.protobuf.Timestamp detectionDateEnd = 3 [(required) = true];
```

**Review by EvanHahn — APPROVED** [review]
LGTM other than one small nitpick.

**Pull request merged**